{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0b508fef-1aef-429d-9dc9-a238bbf78f41",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq  # Import the Seq class\n",
    "from Bio.SeqRecord import SeqRecord  # Import the SeqRecord class for creating sequence records\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "93834c2e-0795-40ad-8a2a-348b4b6cfa18",
   "metadata": {},
   "outputs": [],
   "source": [
    "proj_dir=\"/master/nplatt/sch_hae_its-nigeria\"\n",
    "results_dir=f\"{proj_dir}/results\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "2b1f4c80-0f39-4d7b-b19b-bccf5dd84abc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wgs_id</th>\n",
       "      <th>its_id</th>\n",
       "      <th>sra</th>\n",
       "      <th>species</th>\n",
       "      <th>country</th>\n",
       "      <th>locale</th>\n",
       "      <th>patient</th>\n",
       "      <th>miracidum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sb_NG_ak_1.1</td>\n",
       "      <td>Sb.ng.ak.1.1F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sb_NG_ak_2.1</td>\n",
       "      <td>sb_ng_ak_2.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sb_NG_ak_2.2</td>\n",
       "      <td>Sb.ng.ak.2.2F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sb_NG_ak_2.3</td>\n",
       "      <td>Sb.ng.ak.2.3F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sb_NG_ak_3.1</td>\n",
       "      <td>Sb.ng.ak.3.1R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>Sh_NG_os_3_1</td>\n",
       "      <td>sh_ng_os_3_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>osun</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>c_Sh_NG_os_3_11</td>\n",
       "      <td>Sh.ng.os.3.11F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>osun</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>c_Sh_NG_os_3_5</td>\n",
       "      <td>Sh.ng.os.3.5F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>osun</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>c_Sh_NG_os_3_6</td>\n",
       "      <td>Sh.ng.os.3.6F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>osun</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>SRR11907410</td>\n",
       "      <td>se_sha_pemba_uwandani_13</td>\n",
       "      <td>SRR11907410</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>zanzibar</td>\n",
       "      <td>uwandani</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>205 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              wgs_id                    its_id          sra       species  \\\n",
       "0       Sb_NG_ak_1.1             Sb.ng.ak.1.1F          NaN        sbovis   \n",
       "1       Sb_NG_ak_2.1              sb_ng_ak_2.1          NaN        sbovis   \n",
       "2       Sb_NG_ak_2.2             Sb.ng.ak.2.2F          NaN        sbovis   \n",
       "3       Sb_NG_ak_2.3             Sb.ng.ak.2.3F          NaN        sbovis   \n",
       "4       Sb_NG_ak_3.1             Sb.ng.ak.3.1R          NaN        sbovis   \n",
       "..               ...                       ...          ...           ...   \n",
       "200     Sh_NG_os_3_1              sh_ng_os_3_1          NaN  shaematobium   \n",
       "201  c_Sh_NG_os_3_11            Sh.ng.os.3.11F          NaN  shaematobium   \n",
       "202   c_Sh_NG_os_3_5             Sh.ng.os.3.5F          NaN  shaematobium   \n",
       "203   c_Sh_NG_os_3_6             Sh.ng.os.3.6F          NaN  shaematobium   \n",
       "204      SRR11907410  se_sha_pemba_uwandani_13  SRR11907410  shaematobium   \n",
       "\n",
       "      country    locale patient miracidum  \n",
       "0     nigeria        ak       1         1  \n",
       "1     nigeria        ak       2         1  \n",
       "2     nigeria        ak       2         2  \n",
       "3     nigeria        ak       2         3  \n",
       "4     nigeria        ak       3         1  \n",
       "..        ...       ...     ...       ...  \n",
       "200   nigeria      osun       3         1  \n",
       "201   nigeria      osun       3        11  \n",
       "202   nigeria      osun       3         5  \n",
       "203   nigeria      osun       3         6  \n",
       "204  zanzibar  uwandani      na        na  \n",
       "\n",
       "[205 rows x 8 columns]"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#you can get this table from the SCAN dryad link\n",
    "info_df=pd.read_csv(f\"{proj_dir}/its-nigeria_samplesheet.csv\", sep=\",\")\n",
    "info_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af416e24-eb5f-413e-8c07-a8f9b4e241ff",
   "metadata": {},
   "source": [
    "# Get results from other analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "a624c770-a7fe-4c94-a217-d6fa7cbbe48c",
   "metadata": {},
   "outputs": [],
   "source": [
    "Path(f\"{results_dir}/sanger\").mkdir(parents=True, exist_ok=True)\n",
    "os.chdir(f\"{results_dir}/sanger\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86ac7e21-2966-48a6-bc70-aabbfb273ea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "cp ~/sch_hae_its-nigeria/data/sanger_its_sequences.fas ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c20c8040-a844-4483-85e9-50ef7362d1dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "\n",
    "conda run -n muscle --align sanger_its_sequences.fas --output sanger_its_seqeunces.muscle.fas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "b6bcdae7-a128-48a4-a9fe-be1f3b5a3d9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#manually trimmed the ends so that each sample contains a complete sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "79856e47-33fa-47a8-b57a-eaad8177ea42",
   "metadata": {},
   "outputs": [],
   "source": [
    "its_fasta_file = 'sanger_its_seqeunces.muscle.trimmed.fas'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "9c70760a-0163-4575-94fa-e117bfac02c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df=info_df.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fbd1d50-d421-493c-8b84-520fc4cf89c0",
   "metadata": {},
   "source": [
    "## Add ITS sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "2cbffe72-76da-4b60-b052-f591e44f5277",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wgs_id</th>\n",
       "      <th>its_id</th>\n",
       "      <th>sra</th>\n",
       "      <th>species</th>\n",
       "      <th>country</th>\n",
       "      <th>locale</th>\n",
       "      <th>patient</th>\n",
       "      <th>miracidum</th>\n",
       "      <th>its_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sb_NG_ak_1.1</td>\n",
       "      <td>Sb.ng.ak.1.1F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sb_NG_ak_2.1</td>\n",
       "      <td>sb_ng_ak_2.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sb_NG_ak_2.2</td>\n",
       "      <td>Sb.ng.ak.2.2F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sb_NG_ak_2.3</td>\n",
       "      <td>Sb.ng.ak.2.3F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sb_NG_ak_3.1</td>\n",
       "      <td>Sb.ng.ak.3.1R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>Sh_NG_os_3_1</td>\n",
       "      <td>sh_ng_os_3_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>osun</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>CATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>c_Sh_NG_os_3_11</td>\n",
       "      <td>Sh.ng.os.3.11F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>osun</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>c_Sh_NG_os_3_5</td>\n",
       "      <td>Sh.ng.os.3.5F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>osun</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>c_Sh_NG_os_3_6</td>\n",
       "      <td>Sh.ng.os.3.6F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>osun</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>SRR11907410</td>\n",
       "      <td>se_sha_pemba_uwandani_13</td>\n",
       "      <td>SRR11907410</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>zanzibar</td>\n",
       "      <td>uwandani</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>205 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              wgs_id                    its_id          sra       species  \\\n",
       "0       Sb_NG_ak_1.1             Sb.ng.ak.1.1F          NaN        sbovis   \n",
       "1       Sb_NG_ak_2.1              sb_ng_ak_2.1          NaN        sbovis   \n",
       "2       Sb_NG_ak_2.2             Sb.ng.ak.2.2F          NaN        sbovis   \n",
       "3       Sb_NG_ak_2.3             Sb.ng.ak.2.3F          NaN        sbovis   \n",
       "4       Sb_NG_ak_3.1             Sb.ng.ak.3.1R          NaN        sbovis   \n",
       "..               ...                       ...          ...           ...   \n",
       "200     Sh_NG_os_3_1              sh_ng_os_3_1          NaN  shaematobium   \n",
       "201  c_Sh_NG_os_3_11            Sh.ng.os.3.11F          NaN  shaematobium   \n",
       "202   c_Sh_NG_os_3_5             Sh.ng.os.3.5F          NaN  shaematobium   \n",
       "203   c_Sh_NG_os_3_6             Sh.ng.os.3.6F          NaN  shaematobium   \n",
       "204      SRR11907410  se_sha_pemba_uwandani_13  SRR11907410  shaematobium   \n",
       "\n",
       "      country    locale patient miracidum  \\\n",
       "0     nigeria        ak       1         1   \n",
       "1     nigeria        ak       2         1   \n",
       "2     nigeria        ak       2         2   \n",
       "3     nigeria        ak       2         3   \n",
       "4     nigeria        ak       3         1   \n",
       "..        ...       ...     ...       ...   \n",
       "200   nigeria      osun       3         1   \n",
       "201   nigeria      osun       3        11   \n",
       "202   nigeria      osun       3         5   \n",
       "203   nigeria      osun       3         6   \n",
       "204  zanzibar  uwandani      na        na   \n",
       "\n",
       "                                               its_seq  \n",
       "0    TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...  \n",
       "1    TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...  \n",
       "2    TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...  \n",
       "3    TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...  \n",
       "4    TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...  \n",
       "..                                                 ...  \n",
       "200  CATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...  \n",
       "201  TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...  \n",
       "202  TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...  \n",
       "203  TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...  \n",
       "204  TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...  \n",
       "\n",
       "[205 rows x 9 columns]"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Dictionary to store sequences and their identifiers\n",
    "# Key: sequence (as a string), Value: list of sequence identifiers\n",
    "sequence_dict = defaultdict(list)\n",
    "\n",
    "# Read the FASTA file\n",
    "for record in SeqIO.parse(its_fasta_file, \"fasta\"):\n",
    "    # Append the record.id to the list of ids for this sequence\n",
    "    sequence_dict[record.id].append(str(record.seq))\n",
    "\n",
    "its_sequences=[]\n",
    "\n",
    "for index, row in merged_df.iterrows():\n",
    "    try:\n",
    "        seq=sequence_dict[row[\"its_id\"]][0]\n",
    "        its_sequences.append(seq)\n",
    "    except:\n",
    "        its_sequences.append(\"\")\n",
    "        \n",
    "merged_df[\"its_seq\"] = its_sequences\n",
    "\n",
    "merged_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21052ba3-84f9-41b3-95da-4730a0457180",
   "metadata": {},
   "source": [
    "## Assign ITS seqeunces haplotype IDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "25850ef8-551b-4baf-a1b0-ed0958010285",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wgs_id</th>\n",
       "      <th>its_id</th>\n",
       "      <th>sra</th>\n",
       "      <th>species</th>\n",
       "      <th>country</th>\n",
       "      <th>locale</th>\n",
       "      <th>patient</th>\n",
       "      <th>miracidum</th>\n",
       "      <th>its_seq</th>\n",
       "      <th>genotype_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sb_NG_ak_1.1</td>\n",
       "      <td>Sb.ng.ak.1.1F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...</td>\n",
       "      <td>its-32_n34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sb_NG_ak_2.1</td>\n",
       "      <td>sb_ng_ak_2.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...</td>\n",
       "      <td>its-32_n34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sb_NG_ak_2.2</td>\n",
       "      <td>Sb.ng.ak.2.2F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...</td>\n",
       "      <td>its-32_n34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sb_NG_ak_2.3</td>\n",
       "      <td>Sb.ng.ak.2.3F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...</td>\n",
       "      <td>its-33_n1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sb_NG_ak_3.1</td>\n",
       "      <td>Sb.ng.ak.3.1R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...</td>\n",
       "      <td>its-32_n34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>Sh_NG_os_3_1</td>\n",
       "      <td>sh_ng_os_3_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>osun</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>CATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...</td>\n",
       "      <td>its-3_n7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>c_Sh_NG_os_3_11</td>\n",
       "      <td>Sh.ng.os.3.11F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>osun</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...</td>\n",
       "      <td>its-9_n44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>c_Sh_NG_os_3_5</td>\n",
       "      <td>Sh.ng.os.3.5F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>osun</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...</td>\n",
       "      <td>its-17_n14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>c_Sh_NG_os_3_6</td>\n",
       "      <td>Sh.ng.os.3.6F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>osun</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...</td>\n",
       "      <td>its-9_n44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>SRR11907410</td>\n",
       "      <td>se_sha_pemba_uwandani_13</td>\n",
       "      <td>SRR11907410</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>zanzibar</td>\n",
       "      <td>uwandani</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...</td>\n",
       "      <td>its-9_n44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>205 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              wgs_id                    its_id          sra       species  \\\n",
       "0       Sb_NG_ak_1.1             Sb.ng.ak.1.1F          NaN        sbovis   \n",
       "1       Sb_NG_ak_2.1              sb_ng_ak_2.1          NaN        sbovis   \n",
       "2       Sb_NG_ak_2.2             Sb.ng.ak.2.2F          NaN        sbovis   \n",
       "3       Sb_NG_ak_2.3             Sb.ng.ak.2.3F          NaN        sbovis   \n",
       "4       Sb_NG_ak_3.1             Sb.ng.ak.3.1R          NaN        sbovis   \n",
       "..               ...                       ...          ...           ...   \n",
       "200     Sh_NG_os_3_1              sh_ng_os_3_1          NaN  shaematobium   \n",
       "201  c_Sh_NG_os_3_11            Sh.ng.os.3.11F          NaN  shaematobium   \n",
       "202   c_Sh_NG_os_3_5             Sh.ng.os.3.5F          NaN  shaematobium   \n",
       "203   c_Sh_NG_os_3_6             Sh.ng.os.3.6F          NaN  shaematobium   \n",
       "204      SRR11907410  se_sha_pemba_uwandani_13  SRR11907410  shaematobium   \n",
       "\n",
       "      country    locale patient miracidum  \\\n",
       "0     nigeria        ak       1         1   \n",
       "1     nigeria        ak       2         1   \n",
       "2     nigeria        ak       2         2   \n",
       "3     nigeria        ak       2         3   \n",
       "4     nigeria        ak       3         1   \n",
       "..        ...       ...     ...       ...   \n",
       "200   nigeria      osun       3         1   \n",
       "201   nigeria      osun       3        11   \n",
       "202   nigeria      osun       3         5   \n",
       "203   nigeria      osun       3         6   \n",
       "204  zanzibar  uwandani      na        na   \n",
       "\n",
       "                                               its_seq genotype_label  \n",
       "0    TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...     its-32_n34  \n",
       "1    TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...     its-32_n34  \n",
       "2    TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...     its-32_n34  \n",
       "3    TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...      its-33_n1  \n",
       "4    TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...     its-32_n34  \n",
       "..                                                 ...            ...  \n",
       "200  CATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...       its-3_n7  \n",
       "201  TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...      its-9_n44  \n",
       "202  TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...     its-17_n14  \n",
       "203  TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...      its-9_n44  \n",
       "204  TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...      its-9_n44  \n",
       "\n",
       "[205 rows x 10 columns]"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_groups = (\n",
    "    merged_df.groupby(\"its_seq\")\n",
    "    .agg(genotype_count=(\"its_seq\", \"count\"))\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "genotype_groups\n",
    "\n",
    "# Assign haplotype labels\n",
    "genotype_groups[\"genotype_label\"] = genotype_groups.apply(\n",
    "    lambda row: f\"its-{row.name + 1}_n{row.genotype_count}\", axis=1\n",
    ")\n",
    "\n",
    "# Merge the haplotype labels back to the original DataFrame\n",
    "merged_df = merged_df.merge(\n",
    "    genotype_groups[[\"its_seq\", \"genotype_label\"]],\n",
    "    on=\"its_seq\",\n",
    "    how=\"left\"\n",
    ")\n",
    "\n",
    "# Display the resulting DataFrame\n",
    "merged_df.to_csv(\"its_df.csv\", sep=\",\", header=True, index=False)\n",
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "27204816-0db7-44d9-a600-34d21a5a876f",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"its_genotypes.fas\", 'w') as f:\n",
    "    for index, row in merged_df.iterrows():\n",
    "        its_id=row[\"its_id\"]\n",
    "        geno_id=row[\"genotype_label\"]\n",
    "        seq=row[\"its_seq\"]\n",
    "        f.write(f\">{its_id}#{geno_id}\\n{seq}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "568e01d4-d415-456e-910b-30d8b5957579",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wgs_id</th>\n",
       "      <th>its_id</th>\n",
       "      <th>sra</th>\n",
       "      <th>species</th>\n",
       "      <th>country</th>\n",
       "      <th>locale</th>\n",
       "      <th>patient</th>\n",
       "      <th>miracidum</th>\n",
       "      <th>its_seq</th>\n",
       "      <th>genotype_label</th>\n",
       "      <th>diagnostic_genotype</th>\n",
       "      <th>diagnostic_site_33</th>\n",
       "      <th>diagnostic_site_685</th>\n",
       "      <th>diagnostic_site_740</th>\n",
       "      <th>diagnostic_site_790</th>\n",
       "      <th>diagnostic_site_860</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sb_NG_ak_1.1</td>\n",
       "      <td>Sb.ng.ak.1.1F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...</td>\n",
       "      <td>its-32_n34</td>\n",
       "      <td>GATAT</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sb_NG_ak_2.1</td>\n",
       "      <td>sb_ng_ak_2.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...</td>\n",
       "      <td>its-32_n34</td>\n",
       "      <td>GATAT</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sb_NG_ak_2.2</td>\n",
       "      <td>Sb.ng.ak.2.2F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...</td>\n",
       "      <td>its-32_n34</td>\n",
       "      <td>GATAT</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sb_NG_ak_2.3</td>\n",
       "      <td>Sb.ng.ak.2.3F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...</td>\n",
       "      <td>its-33_n1</td>\n",
       "      <td>GATAT</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sb_NG_ak_3.1</td>\n",
       "      <td>Sb.ng.ak.3.1R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...</td>\n",
       "      <td>its-32_n34</td>\n",
       "      <td>GATAT</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>Sh_NG_os_3_1</td>\n",
       "      <td>sh_ng_os_3_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>osun</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>CATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...</td>\n",
       "      <td>its-3_n7</td>\n",
       "      <td>AGCGC</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>c_Sh_NG_os_3_11</td>\n",
       "      <td>Sh.ng.os.3.11F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>osun</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...</td>\n",
       "      <td>its-9_n44</td>\n",
       "      <td>AGCGC</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>c_Sh_NG_os_3_5</td>\n",
       "      <td>Sh.ng.os.3.5F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>osun</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...</td>\n",
       "      <td>its-17_n14</td>\n",
       "      <td>AGCGC</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>c_Sh_NG_os_3_6</td>\n",
       "      <td>Sh.ng.os.3.6F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>osun</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...</td>\n",
       "      <td>its-9_n44</td>\n",
       "      <td>AGCGC</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>SRR11907410</td>\n",
       "      <td>se_sha_pemba_uwandani_13</td>\n",
       "      <td>SRR11907410</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>zanzibar</td>\n",
       "      <td>uwandani</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...</td>\n",
       "      <td>its-9_n44</td>\n",
       "      <td>AGCGC</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>205 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              wgs_id                    its_id          sra       species  \\\n",
       "0       Sb_NG_ak_1.1             Sb.ng.ak.1.1F          NaN        sbovis   \n",
       "1       Sb_NG_ak_2.1              sb_ng_ak_2.1          NaN        sbovis   \n",
       "2       Sb_NG_ak_2.2             Sb.ng.ak.2.2F          NaN        sbovis   \n",
       "3       Sb_NG_ak_2.3             Sb.ng.ak.2.3F          NaN        sbovis   \n",
       "4       Sb_NG_ak_3.1             Sb.ng.ak.3.1R          NaN        sbovis   \n",
       "..               ...                       ...          ...           ...   \n",
       "200     Sh_NG_os_3_1              sh_ng_os_3_1          NaN  shaematobium   \n",
       "201  c_Sh_NG_os_3_11            Sh.ng.os.3.11F          NaN  shaematobium   \n",
       "202   c_Sh_NG_os_3_5             Sh.ng.os.3.5F          NaN  shaematobium   \n",
       "203   c_Sh_NG_os_3_6             Sh.ng.os.3.6F          NaN  shaematobium   \n",
       "204      SRR11907410  se_sha_pemba_uwandani_13  SRR11907410  shaematobium   \n",
       "\n",
       "      country    locale patient miracidum  \\\n",
       "0     nigeria        ak       1         1   \n",
       "1     nigeria        ak       2         1   \n",
       "2     nigeria        ak       2         2   \n",
       "3     nigeria        ak       2         3   \n",
       "4     nigeria        ak       3         1   \n",
       "..        ...       ...     ...       ...   \n",
       "200   nigeria      osun       3         1   \n",
       "201   nigeria      osun       3        11   \n",
       "202   nigeria      osun       3         5   \n",
       "203   nigeria      osun       3         6   \n",
       "204  zanzibar  uwandani      na        na   \n",
       "\n",
       "                                               its_seq genotype_label  \\\n",
       "0    TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...     its-32_n34   \n",
       "1    TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...     its-32_n34   \n",
       "2    TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...     its-32_n34   \n",
       "3    TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...      its-33_n1   \n",
       "4    TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...     its-32_n34   \n",
       "..                                                 ...            ...   \n",
       "200  CATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...       its-3_n7   \n",
       "201  TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...      its-9_n44   \n",
       "202  TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...     its-17_n14   \n",
       "203  TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...      its-9_n44   \n",
       "204  TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...      its-9_n44   \n",
       "\n",
       "    diagnostic_genotype diagnostic_site_33 diagnostic_site_685  \\\n",
       "0                 GATAT                  G                   A   \n",
       "1                 GATAT                  G                   A   \n",
       "2                 GATAT                  G                   A   \n",
       "3                 GATAT                  G                   A   \n",
       "4                 GATAT                  G                   A   \n",
       "..                  ...                ...                 ...   \n",
       "200               AGCGC                  A                   G   \n",
       "201               AGCGC                  A                   G   \n",
       "202               AGCGC                  A                   G   \n",
       "203               AGCGC                  A                   G   \n",
       "204               AGCGC                  A                   G   \n",
       "\n",
       "    diagnostic_site_740 diagnostic_site_790 diagnostic_site_860  \n",
       "0                     T                   A                   T  \n",
       "1                     T                   A                   T  \n",
       "2                     T                   A                   T  \n",
       "3                     T                   A                   T  \n",
       "4                     T                   A                   T  \n",
       "..                  ...                 ...                 ...  \n",
       "200                   C                   G                   C  \n",
       "201                   C                   G                   C  \n",
       "202                   C                   G                   C  \n",
       "203                   C                   G                   C  \n",
       "204                   C                   G                   C  \n",
       "\n",
       "[205 rows x 16 columns]"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diagnostic_positions = [33, 685, 740, 790, 860]\n",
    "\n",
    "sample_diagnostic_calls = []\n",
    "sample_diagnostic_genotype=[]\n",
    "\n",
    "for index, row in merged_df.iterrows():\n",
    "    seq=row[\"its_seq\"]\n",
    "    try:\n",
    "        calls=[seq[pos - 1] for pos in diagnostic_positions]\n",
    "        genotype=\"\".join(calls)\n",
    "        \n",
    "        sample_diagnostic_calls.append(calls)\n",
    "        sample_diagnostic_genotype.append(genotype)\n",
    "    except:\n",
    "        calls=[\"?\",\"?\",\"?\",\"?\",\"?\"]\n",
    "        genotype=\"\".join(calls)\n",
    "    \n",
    "        sample_diagnostic_calls.append(calls)\n",
    "        sample_diagnostic_genotype.append(genotype)\n",
    "\n",
    "merged_df[\"diagnostic_genotype\"]=sample_diagnostic_genotype\n",
    "\n",
    "# Create a DataFrame from the diagnostic calls\n",
    "diagnostic_df = pd.DataFrame(\n",
    "    sample_diagnostic_calls, \n",
    "    columns=[f\"diagnostic_site_{pos}\" for pos in diagnostic_positions]\n",
    ")\n",
    "\n",
    "# Concatenate the diagnostic DataFrame with merged_df\n",
    "merged_df = pd.concat([merged_df, diagnostic_df], axis=1)\n",
    "\n",
    "# Display the updated DataFrame\n",
    "merged_df.to_csv(\"its_df.csv\", sep=\",\", header=True, index=False)\n",
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "0e68c306-9dd6-4f69-b97b-e2bd4a61d2fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['GATAT', 'AGCGC', 'ARYRC', 'ARYRY', 'AATAT', 'ARYAT', 'ARCRC',\n",
       "       '?????'], dtype=object)"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df[\"diagnostic_genotype\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "e8620a83-4436-4b3b-972c-865637ee1b40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wgs_id</th>\n",
       "      <th>its_id</th>\n",
       "      <th>sra</th>\n",
       "      <th>species</th>\n",
       "      <th>country</th>\n",
       "      <th>locale</th>\n",
       "      <th>patient</th>\n",
       "      <th>miracidum</th>\n",
       "      <th>its_seq</th>\n",
       "      <th>genotype_label</th>\n",
       "      <th>diagnostic_genotype</th>\n",
       "      <th>diagnostic_site_33</th>\n",
       "      <th>diagnostic_site_685</th>\n",
       "      <th>diagnostic_site_740</th>\n",
       "      <th>diagnostic_site_790</th>\n",
       "      <th>diagnostic_site_860</th>\n",
       "      <th>its_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sb_NG_ak_1.1</td>\n",
       "      <td>Sb.ng.ak.1.1F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...</td>\n",
       "      <td>its-32_n34</td>\n",
       "      <td>GATAT</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>SBxSB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sb_NG_ak_2.1</td>\n",
       "      <td>sb_ng_ak_2.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...</td>\n",
       "      <td>its-32_n34</td>\n",
       "      <td>GATAT</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>SBxSB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sb_NG_ak_2.2</td>\n",
       "      <td>Sb.ng.ak.2.2F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...</td>\n",
       "      <td>its-32_n34</td>\n",
       "      <td>GATAT</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>SBxSB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sb_NG_ak_2.3</td>\n",
       "      <td>Sb.ng.ak.2.3F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...</td>\n",
       "      <td>its-33_n1</td>\n",
       "      <td>GATAT</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>SBxSB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sb_NG_ak_3.1</td>\n",
       "      <td>Sb.ng.ak.3.1R</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sbovis</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>ak</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...</td>\n",
       "      <td>its-32_n34</td>\n",
       "      <td>GATAT</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>SBxSB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>Sh_NG_os_3_1</td>\n",
       "      <td>sh_ng_os_3_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>osun</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>CATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...</td>\n",
       "      <td>its-3_n7</td>\n",
       "      <td>AGCGC</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>SHxSH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>c_Sh_NG_os_3_11</td>\n",
       "      <td>Sh.ng.os.3.11F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>osun</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...</td>\n",
       "      <td>its-9_n44</td>\n",
       "      <td>AGCGC</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>SHxSH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>c_Sh_NG_os_3_5</td>\n",
       "      <td>Sh.ng.os.3.5F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>osun</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...</td>\n",
       "      <td>its-17_n14</td>\n",
       "      <td>AGCGC</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>SHxSH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>c_Sh_NG_os_3_6</td>\n",
       "      <td>Sh.ng.os.3.6F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>nigeria</td>\n",
       "      <td>osun</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...</td>\n",
       "      <td>its-9_n44</td>\n",
       "      <td>AGCGC</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>SHxSH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>SRR11907410</td>\n",
       "      <td>se_sha_pemba_uwandani_13</td>\n",
       "      <td>SRR11907410</td>\n",
       "      <td>shaematobium</td>\n",
       "      <td>zanzibar</td>\n",
       "      <td>uwandani</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...</td>\n",
       "      <td>its-9_n44</td>\n",
       "      <td>AGCGC</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>SHxSH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>205 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              wgs_id                    its_id          sra       species  \\\n",
       "0       Sb_NG_ak_1.1             Sb.ng.ak.1.1F          NaN        sbovis   \n",
       "1       Sb_NG_ak_2.1              sb_ng_ak_2.1          NaN        sbovis   \n",
       "2       Sb_NG_ak_2.2             Sb.ng.ak.2.2F          NaN        sbovis   \n",
       "3       Sb_NG_ak_2.3             Sb.ng.ak.2.3F          NaN        sbovis   \n",
       "4       Sb_NG_ak_3.1             Sb.ng.ak.3.1R          NaN        sbovis   \n",
       "..               ...                       ...          ...           ...   \n",
       "200     Sh_NG_os_3_1              sh_ng_os_3_1          NaN  shaematobium   \n",
       "201  c_Sh_NG_os_3_11            Sh.ng.os.3.11F          NaN  shaematobium   \n",
       "202   c_Sh_NG_os_3_5             Sh.ng.os.3.5F          NaN  shaematobium   \n",
       "203   c_Sh_NG_os_3_6             Sh.ng.os.3.6F          NaN  shaematobium   \n",
       "204      SRR11907410  se_sha_pemba_uwandani_13  SRR11907410  shaematobium   \n",
       "\n",
       "      country    locale patient miracidum  \\\n",
       "0     nigeria        ak       1         1   \n",
       "1     nigeria        ak       2         1   \n",
       "2     nigeria        ak       2         2   \n",
       "3     nigeria        ak       2         3   \n",
       "4     nigeria        ak       3         1   \n",
       "..        ...       ...     ...       ...   \n",
       "200   nigeria      osun       3         1   \n",
       "201   nigeria      osun       3        11   \n",
       "202   nigeria      osun       3         5   \n",
       "203   nigeria      osun       3         6   \n",
       "204  zanzibar  uwandani      na        na   \n",
       "\n",
       "                                               its_seq genotype_label  \\\n",
       "0    TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...     its-32_n34   \n",
       "1    TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...     its-32_n34   \n",
       "2    TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...     its-32_n34   \n",
       "3    TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...      its-33_n1   \n",
       "4    TATAATGATGCATGCACCTGGCTTCTTGCTGGGCTGTATGTACCCT...     its-32_n34   \n",
       "..                                                 ...            ...   \n",
       "200  CATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...       its-3_n7   \n",
       "201  TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...      its-9_n44   \n",
       "202  TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...     its-17_n14   \n",
       "203  TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...      its-9_n44   \n",
       "204  TATAATGATGCATGCACCTGGCTTCTTGCTGGACTGTATGTACCCT...      its-9_n44   \n",
       "\n",
       "    diagnostic_genotype diagnostic_site_33 diagnostic_site_685  \\\n",
       "0                 GATAT                  G                   A   \n",
       "1                 GATAT                  G                   A   \n",
       "2                 GATAT                  G                   A   \n",
       "3                 GATAT                  G                   A   \n",
       "4                 GATAT                  G                   A   \n",
       "..                  ...                ...                 ...   \n",
       "200               AGCGC                  A                   G   \n",
       "201               AGCGC                  A                   G   \n",
       "202               AGCGC                  A                   G   \n",
       "203               AGCGC                  A                   G   \n",
       "204               AGCGC                  A                   G   \n",
       "\n",
       "    diagnostic_site_740 diagnostic_site_790 diagnostic_site_860 its_class  \n",
       "0                     T                   A                   T     SBxSB  \n",
       "1                     T                   A                   T     SBxSB  \n",
       "2                     T                   A                   T     SBxSB  \n",
       "3                     T                   A                   T     SBxSB  \n",
       "4                     T                   A                   T     SBxSB  \n",
       "..                  ...                 ...                 ...       ...  \n",
       "200                   C                   G                   C     SHxSH  \n",
       "201                   C                   G                   C     SHxSH  \n",
       "202                   C                   G                   C     SHxSH  \n",
       "203                   C                   G                   C     SHxSH  \n",
       "204                   C                   G                   C     SHxSH  \n",
       "\n",
       "[205 rows x 17 columns]"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diagnostic_classes = {'AGCGC': 'SHxSH', \n",
    "                      'ARYRC': 'SHxSC', \n",
    "                      'ARYRY': 'SHxSC', \n",
    "                      'AATAT': 'SCxSC', \n",
    "                      'GATAT': 'SBxSB', \n",
    "                      'ARCRC': 'SHxSC',\n",
    "                      'ARYAT': 'SHxSC',\n",
    "                      '?????': 'NA'}\n",
    "\n",
    "\n",
    "classes=[]\n",
    "for index, row in merged_df.iterrows():\n",
    "    classes.append(diagnostic_classes[row[\"diagnostic_genotype\"]])\n",
    "\n",
    "merged_df[\"its_class\"] = classes\n",
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "d66194d9-adfd-4868-be69-8d211cb26984",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df[[\"wgs_id\", \"its_id\", \"its_seq\", \"genotype_label\", \"diagnostic_genotype\", \"diagnostic_site_33\", \"diagnostic_site_685\", \"diagnostic_site_740\", \n",
    "           \"diagnostic_site_790\", \"diagnostic_site_860\", \"its_class\"]].to_csv(\"its_class.csv\", sep=\",\", header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "119601d3-1132-4da4-90d3-a332f9dc97d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wgs_id</th>\n",
       "      <th>its_id</th>\n",
       "      <th>sra</th>\n",
       "      <th>species</th>\n",
       "      <th>country</th>\n",
       "      <th>locale</th>\n",
       "      <th>patient</th>\n",
       "      <th>miracidum</th>\n",
       "      <th>its_seq</th>\n",
       "      <th>genotype_label</th>\n",
       "      <th>diagnostic_genotype</th>\n",
       "      <th>diagnostic_site_33</th>\n",
       "      <th>diagnostic_site_685</th>\n",
       "      <th>diagnostic_site_740</th>\n",
       "      <th>diagnostic_site_790</th>\n",
       "      <th>diagnostic_site_860</th>\n",
       "      <th>its_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>ERR119612</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERR119612</td>\n",
       "      <td>guineensis</td>\n",
       "      <td>saotome</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td></td>\n",
       "      <td>its-1_n26</td>\n",
       "      <td>?????</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>ERR119613</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERR119613</td>\n",
       "      <td>intercalatum</td>\n",
       "      <td>drcongo</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td></td>\n",
       "      <td>its-1_n26</td>\n",
       "      <td>?????</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>ERR310940</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERR310940</td>\n",
       "      <td>margrebowiei</td>\n",
       "      <td>zambia</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td></td>\n",
       "      <td>its-1_n26</td>\n",
       "      <td>?????</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>ERR103051</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERR103051</td>\n",
       "      <td>matthei</td>\n",
       "      <td>zambia</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td></td>\n",
       "      <td>its-1_n26</td>\n",
       "      <td>?????</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        wgs_id its_id        sra       species  country locale patient  \\\n",
       "148  ERR119612    NaN  ERR119612    guineensis  saotome     na      na   \n",
       "149  ERR119613    NaN  ERR119613  intercalatum  drcongo     na      na   \n",
       "150  ERR310940    NaN  ERR310940  margrebowiei   zambia     na      na   \n",
       "151  ERR103051    NaN  ERR103051       matthei   zambia     na      na   \n",
       "\n",
       "    miracidum its_seq genotype_label diagnostic_genotype diagnostic_site_33  \\\n",
       "148        na              its-1_n26               ?????                  ?   \n",
       "149        na              its-1_n26               ?????                  ?   \n",
       "150        na              its-1_n26               ?????                  ?   \n",
       "151        na              its-1_n26               ?????                  ?   \n",
       "\n",
       "    diagnostic_site_685 diagnostic_site_740 diagnostic_site_790  \\\n",
       "148                   ?                   ?                   ?   \n",
       "149                   ?                   ?                   ?   \n",
       "150                   ?                   ?                   ?   \n",
       "151                   ?                   ?                   ?   \n",
       "\n",
       "    diagnostic_site_860 its_class  \n",
       "148                   ?        NA  \n",
       "149                   ?        NA  \n",
       "150                   ?        NA  \n",
       "151                   ?        NA  "
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_species = ['shaematobium', 'sbovis', 'shxsb', 'sbxsc', 'scurassoni']\n",
    "merged_df.loc[~merged_df[\"species\"].isin(target_species)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "9bc1024a-ff78-452d-b2ef-3ba5324d34d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "its_class\n",
       "SHxSH    72\n",
       "SBxSB    41\n",
       "SHxSC    37\n",
       "SCxSC    29\n",
       "NA       26\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df[\"its_class\"].value_counts()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "popgen",
   "language": "python",
   "name": "popgen"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
